File: mods/java/hytale/src/main/java/com/alaydriem/bedrockvoicechat/hytale/HytalePlugin.java
package com.alaydriem.bedrockvoicechat.hytale;

import com.alaydriem.bedrockvoicechat.config.ModConfig;
import com.alaydriem.bedrockvoicechat.dto.Payload;
import com.alaydriem.bedrockvoicechat.dto.PlayerData;
import com.alaydriem.bedrockvoicechat.network.HttpRequestHandler;

import com.hypixel.hytale.server.core.HytaleServer;
import com.hypixel.hytale.server.core.plugin.JavaPlugin;
import com.hypixel.hytale.server.core.plugin.JavaPluginInit;
import com.hypixel.hytale.server.core.event.events.player.PlayerConnectEvent;
import com.hypixel.hytale.server.core.event.events.player.PlayerDisconnectEvent;
import com.hypixel.hytale.server.core.util.Config;

import java.util.List;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;

/**
 * Hytale plugin entry point for Bedrock Voice Chat.
 */
public class HytalePlugin extends JavaPlugin {
    private final Config<HytaleModConfig> config;
    private final HytaleConfigProvider configProvider;
    private final HytalePlayerDataProvider playerDataProvider = new HytalePlayerDataProvider();

    private ModConfig modConfig;
    private HttpRequestHandler httpHandler;
    private ScheduledFuture<?> tickTask;

    public HytalePlugin(JavaPluginInit init) {
        super(init);
        // Initialize config using Hytale's withConfig pattern
        this.config = withConfig("bedrock-voice-chat", HytaleModConfig.CODEC);
        this.configProvider = new HytaleConfigProvider(this.config);
    }

    @Override
    protected void setup() {
        getLogger().at(Level.INFO).log("Initializing Bedrock Voice Chat for Hytale");

        // Load config via provider (now backed by Hytale's Config<T>)
        modConfig = configProvider.load();
        if (!modConfig.isValid()) {
            getLogger().at(Level.SEVERE).log("Invalid configuration - plugin will not track players");
            return;
        }

        httpHandler = new HttpRequestHandler(modConfig.getBvcServer(), modConfig.getAccessToken());

        // Register player connect/disconnect events
        getEventRegistry().register(PlayerConnectEvent.class, this::onPlayerConnect);
        getEventRegistry().register(PlayerDisconnectEvent.class, this::onPlayerDisconnect);

        // Schedule tick task using Hytale's built-in executor
        // ~167ms interval = roughly 5 ticks at 30 TPS (Hytale default)
        tickTask = HytaleServer.SCHEDULED_EXECUTOR.scheduleAtFixedRate(
                this::tick, 167, 167, TimeUnit.MILLISECONDS
        );

        getLogger().at(Level.INFO).log("Bedrock Voice Chat will connect to: " + modConfig.getBvcServer());
    }

    @Override
    protected void shutdown() {
        if (tickTask != null && !tickTask.isCancelled()) {
            tickTask.cancel(false);
        }
    }

    private void onPlayerConnect(PlayerConnectEvent event) {
        playerDataProvider.addPlayer(event.getPlayerRef());
    }

    private void onPlayerDisconnect(PlayerDisconnectEvent event) {
        playerDataProvider.removePlayer(event.getPlayerRef());
    }

    private void tick() {
        try {
            List<PlayerData> players = playerDataProvider.collectPlayers();

            if (players.size() < Math.max(modConfig.getMinimumPlayers(), 2)) {
                return;
            }

            Payload payload = new Payload(playerDataProvider.getGameType(), players);
            httpHandler.sendAsync(payload);
        } catch (Exception e) {
            getLogger().at(Level.WARNING).log("Error during tick: " + e.getMessage());
        }
    }
}

File: mods/java/hytale/src/main/kotlin/com/alaydriem/bedrockvoicechat/hytale/HytalePlugin.kt
package com.alaydriem.bedrockvoicechat.hytale

import com.alaydriem.bedrockvoicechat.dto.Payload
import com.alaydriem.bedrockvoicechat.hytale.systems.CrouchTickingSystem
import com.alaydriem.bedrockvoicechat.hytale.systems.DeathChangeSystem
import com.alaydriem.bedrockvoicechat.hytale.systems.SpectatorChangeSystem
import com.alaydriem.bedrockvoicechat.native.PositionSender
import com.alaydriem.bedrockvoicechat.network.HttpRequestHandler
import com.alaydriem.bedrockvoicechat.server.BvcServerManager
import com.hypixel.hytale.server.core.event.events.player.PlayerConnectEvent
import com.hypixel.hytale.server.core.event.events.player.PlayerDisconnectEvent
import com.hypixel.hytale.server.core.plugin.JavaPlugin
import com.hypixel.hytale.server.core.plugin.JavaPluginInit
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.logging.Level

/**
 * Hytale plugin entry point for Bedrock Voice Chat.
 */
class HytalePlugin(init: JavaPluginInit) : JavaPlugin(init) {
    private val config = withConfig("BedrockVoiceChatPlugin", BedrockVoiceChatConfig.CODEC)
    private val configProvider = HytaleConfigProvider(config)
    private val playerDataProvider = HytalePlayerDataProvider()

    private var embeddedServer: BvcServerManager? = null
    private var positionSender: PositionSender? = null
    private var threadPool: ScheduledExecutorService? = null
    private var tickTask: ScheduledFuture<*>? = null
    private var crouchSystem: CrouchTickingSystem? = null
    private var minimumPlayers = 2

    override fun setup() {
        logger.at(Level.INFO).log("Initializing Bedrock Voice Chat for Hytale")

        // Save config to create file if it doesn't exist (with defaults)
        config.save()

        // Load config via provider (backed by Hytale's Config<T>)
        val modConfig = configProvider.load()
        if (!modConfig.isValid()) {
            logger.at(Level.SEVERE).log("Invalid configuration. Bedrock Voice Chat will not be enabled.")
            logger.at(Level.SEVERE).log("Config validation failed: useEmbeddedServer=${modConfig.useEmbeddedServer}, " +
                "bvcServer=${if (modConfig.bvcServer.isNullOrBlank()) "MISSING" else "set"}, " +
                "accessToken=${if (modConfig.accessToken.isNullOrBlank()) "MISSING" else "set"}")
            return
        }

        minimumPlayers = modConfig.minimumPlayers

        // Initialize embedded server if configured
        if (modConfig.useEmbeddedServer) {
            embeddedServer = BvcServerManager(modConfig, configProvider)
            if (!embeddedServer!!.start()) {
                logger.at(Level.SEVERE).log("Failed to start embedded server - falling back to disabled state")
                embeddedServer = null
                return
            }

            // Embedded mode: use FFI directly, no HTTP handler needed
            positionSender = PositionSender(null, embeddedServer)

            val embedded = modConfig.embeddedConfig
            logger.at(Level.INFO).log("Bedrock Voice Chat using embedded server (QUIC port: ${embedded?.quicPort ?: 8443})")
        } else {
            // External server mode: use HTTP handler
            val httpHandler = HttpRequestHandler(modConfig.bvcServer!!, modConfig.accessToken!!)
            positionSender = PositionSender(httpHandler, null)

            logger.at(Level.INFO).log("Bedrock Voice Chat will connect to: ${modConfig.bvcServer}")
        }

        // Create our own thread pool for async tick (isolated from Hytale's internal threading)
        threadPool = Executors.newScheduledThreadPool(2) { r ->
            Thread(r).apply {
                name = "BVC-Thread-${threadId()}"
                isDaemon = true
            }
        }

        // Register player connect/disconnect events
        registerEvents()

        // Register ECS systems for component change detection (runs on world thread)
        registerSystems()

        // Schedule tick on our thread pool (reads from cache only, no component access)
        tickTask = threadPool?.scheduleAtFixedRate(
            { tick() }, 167, 167, TimeUnit.MILLISECONDS
        )

        logger.at(Level.INFO).log("Bedrock Voice Chat initialized successfully")
    }

    override fun shutdown() {
        // Cancel tick task
        tickTask?.let { task ->
            if (!task.isCancelled) {
                task.cancel(false)
            }
        }
        tickTask = null

        // Shutdown thread pool gracefully
        threadPool?.shutdown()
        try {
            if (threadPool?.awaitTermination(5, TimeUnit.SECONDS) != true) {
                threadPool?.shutdownNow()
            }
        } catch (e: InterruptedException) {
            threadPool?.shutdownNow()
        }
        threadPool = null

        embeddedServer?.stop()
        logger.at(Level.INFO).log("Bedrock Voice Chat disabled")
    }

    private fun registerEvents() {
        // Player connect/disconnect for tracking online players
        eventRegistry.register(PlayerConnectEvent::class.java) { event ->
            playerDataProvider.addPlayer(event.playerRef)
        }
        eventRegistry.register(PlayerDisconnectEvent::class.java) { event ->
            playerDataProvider.removePlayer(event.playerRef)
            crouchSystem?.removePlayer(event.playerRef.uuid)
        }
    }

    private fun registerSystems() {
        // Death detection via ECS system (fires when DeathComponent is added/removed)
        val deathSystem = DeathChangeSystem(
            { uuid -> playerDataProvider.markDead(uuid) },
            { uuid -> playerDataProvider.markAlive(uuid) }
        )
        entityStoreRegistry.registerSystem(deathSystem)
        logger.at(Level.INFO).log("[BVC] Registered DeathChangeSystem")

        // Crouch detection via tick-based system (polls MovementStatesComponent every tick)
        val crouchSys = CrouchTickingSystem { uuid, crouching ->
            playerDataProvider.setCrouching(uuid, crouching)
        }
        crouchSystem = crouchSys
        entityStoreRegistry.registerSystem(crouchSys)
        logger.at(Level.INFO).log("[BVC] Registered CrouchTickingSystem")

        // Spectator detection via ECS system (fires when HiddenFromAdventurePlayers is added/removed)
        val spectatorSystem = SpectatorChangeSystem { uuid, spectator ->
            playerDataProvider.setSpectator(uuid, spectator)
        }
        entityStoreRegistry.registerSystem(spectatorSystem)
        logger.at(Level.INFO).log("[BVC] Registered SpectatorChangeSystem")
    }

    private fun tick() {
        try {
            val sender = positionSender ?: return
            val players = playerDataProvider.collectPlayers()

            if (players.size < minimumPlayers) {
                return
            }

            val payload = Payload(playerDataProvider.getGameType(), players)
            sender.send(payload)
        } catch (e: Exception) {
            logger.at(Level.WARNING).log("Error during tick: ${e.message}")
        }
    }
}
